[8.0][IMP] Transfer the Gross Profit fields to the Deliverable lines

sudhir-serpentcs opened the pull request:
Transfer the Gross Profit fields to the Deliverable lines.

BR with BRD_cost give the calculation for Total Revenue-Total Cost = Gross Profit
Need to have on every DL the same calculation based on their own RL
So DL 1 will have a partial Total Revenue DL 1 - Total Cost DL 1 = Gross Profit DL 1



@elicoidal

elicoidal commented on business_requirement_deliverable_cost/models/business.py:
```diff
@@ -123,6 +123,24 @@ class BusinessRequirementDeliverable(models.Model):
+    )
+
+    @api.multi
+    @api.depends('resource_ids.price_total',
```
I am not sure this does what I am expecting.
Here is what I need: DL.gross_profit = DL.price_total - sum(RL.cost)
It should be updated whenever the RL.cost or DL.price_total changes.

Can you double check? @sudhir-serpentcs

elicoidal commented on business_requirement_deliverable_cost/views/business.xml:
```diff
@@ -31,6 +31,9 @@
                     </group>
                 </xpath>
+                <xpath expr="//field[@name='price_total']" position="after">
+                    <field name="gross_profit"/>
```
Not only: I would like to see the same fields as in the BR (check screenshot that I have provided).
Besides it should only be visible to the "cost" group

sudhir-serpentcs commented:
@elicoidal  
Changed calculation for gross profit and added ACL.
https://github.com/OCA/business-requirement/pull/105/commits/a59e162ee32249967115198f7003980f14f06225

elicoidal commented:
@sudhir-serpentcs please check this and the screenshot. The current code doesnot cover it all AFAIK.

elicoidal commented:
please follow: https://github.com/OCA/business-requirement/issues/84

sudhir-serpentcs commented:
@elicoidal @victormartinelicocorp 
can you please review code.

elicoidal commented:
Will need to be fixed when #113 is finished

elicoidal commented:
@sudhir-serpentcs can you rebase your work to fit the changes made in #113

elicoidal commented:
@sudhir-serpentcs once you fix the link to #113 we could quickly merge

elicoidal commented:
@sudhir-serpentcs #113 is fixed. Can you rebase and fix this one?

sudhir-serpentcs commented:
@elicoidal 
please review code.

elicoidal commented:
@sudhir-serpentcs I have lost the description field in deliverable

sudhir-serpentcs commented:
@elicoidal 
Added Description field, please review code.

elicoidal commented:
@sudhir-serpentcs far from been ready: can you retest the full process?

When creating a RL in the DL, I have following js error: Error: NameError: name 'business_requirement_id' is not defined

sudhir-serpentcs commented:
@elicoidal  
Improved code as per comment.
please review code

elicoidal commented:
@sudhir-serpentcs thanks it works now.
I want to change the position of the fields as it is a bit weird as is (so you will have probably to update both modules).

First column:
* Business Requirement
* Product
* Name

Second column:
* Quantity 
* Sales Price
* Currency
* Total revenue
* Total procurement
* Total tasks
* Estimated Gross Profit

sudhir-serpentcs commented:
@elicoidal 
changed the position of the fields.
please review code.

elicoidal commented:
@sudhir-serpentcs Position of the fields is correct in DL form. Thanks!

I have the following small improvements:
- [X]  Total revenue+currency should appear in BR tree view and total at bottom (#108). Should appear in pivot view.
- [X] RL contains price/cost info but not when they are in DL.
- [X] Currency is not set if no customer/pricelist is select. It should be the company currency by default.
- [X] Responsible is not in the BR tree View. It should be the first
- [X] Remove "Created on" in the BR Tree view
- [X] the smart button Deliverable Lines should add the total revenue between parenthesis. Rename Deliverable Lines => Deliverable
- [X] the smart button Resource Lines should add the total cost between parenthesis. Rename Resources Lines => Resources

sudhir-serpentcs commented:
@elicoidal 
Changed code as per comment.
please review code.

elicoidal commented:
@sudhir-serpentcs almost there. A couple of details left:
- [X] RL contains Sales price/Revenue info but not when they are in DL (my previous comment was not clear)
- [X] add the currency in tab Cost control in BR, same as in deliverable.
- [ ] In BR pivot: I dont have the Total revenue, Total Cost and Gross profit as measure
- [ ] In deliverable pivot: I dont have the Total revenue, Total Cost and Gross profit as measure
- [X] In resource pivot: I dont have the Total revenue and Total Cost as measure (While they appear in the table)

Rest looks awesome!

sudhir-serpentcs commented:
@elicoidal 
Improved pivot views & code as per comments.
please review code.

elicoidal commented:
@sudhir-serpentcs still not there (check: https://github.com/OCA/business-requirement/pull/105#issuecomment-297589631):
- [ ] DL and BR pivot are not complete. 
- [ ] BR pivot should be table by default and is broken. Not all measure are available
- [ ] DL Pivot: I miss cost fields (unit price and total)

sudhir-serpentcs commented:
@elicoidal 

>DL Pivot: I miss cost fields (unit price and total)

Changed code as per comments, except the above point as I could not find the mentioned fields.

Please review code.

CC @victormartinelicocorp @darshan-serpent

dreispt commented on business_requirement/views/business_view.xml:
```diff
@@ -107,6 +107,7 @@
             <field name="model">business.requirement</field>
             <field name="arch" type="xml">
                 <tree string="Business Requirement">
+                    <field name="responsible_id"/>
```
Missing one indentation space

dreispt commented on business_requirement/views/business_view.xml:
```diff
@@ -135,7 +135,7 @@
                         <filter string="State" icon="terp-go-home" context="{'group_by':'state'}"/>
                         <filter string="Priority" icon="terp-go-home" context="{'group_by':'priority'}"/>
-                        <filter string="Responsible" icon="terp-go-home" context="{'group_by':'required'}"/>
+                        <filter string="Responsible" icon="terp-go-home" context="{'group_by':'responsible_id'}"/>
```
Icons are deprecated afaik

dreispt commented on business_requirement_deliverable/models/business.py:
```diff
@@ -144,12 +144,13 @@ class BusinessRequirementDeliverable(models.Model):
         string='Currency',
         readonly=True,
-        compute='_compute_get_currency',
+        compute='_compute_get_currency'
```
No problem with leaving the trailing comma here

dreispt commented on business_requirement_deliverable/models/business.py:
```diff
@@ -172,6 +173,8 @@ def _compute_get_currency(self):
             if currency_id:
                 brd.currency_id = currency_id
+            else:
+                brd.currency_id = self.env.user.company_id.currency_id.id
```
Afaict you should remove the trailing `.id`: you want to reference the Currency record, and not it's numeric id.

dreispt commented on business_requirement_deliverable/tests/test_br.py:
```diff
@@ -149,7 +149,13 @@ def test_open_business_requirement_dl_rl(self):
 
+    def test_compute_total_revenue_cost(self):
+        self.br._compute_dl_total_revenue()
+        self.br._compute_rl_total_cost()
```
You're not actually testing the code here.
You should add lines to a BR, and then read the totals and compare with the expected value...

dreispt commented on business_requirement_deliverable_cost/models/business.py:
```diff
@@ -181,7 +227,7 @@ def _compute_resource_task_total(self):
                         'resource_ids').filtered(
                         lambda r: r.resource_type == 'task').mapped(
-                            'price_total'))
+                        'price_total'))
```
The indentation was correct.

dreispt commented on business_requirement_deliverable_cost/models/business.py:
```diff
@@ -192,7 +238,7 @@ def _compute_resource_procurement_total(self):
                         'resource_ids').filtered(
                         lambda r: r.resource_type == 'procurement').mapped(
-                            'price_total'))
+                        'price_total'))
```
Don't remove the indentation.

elicoidal commented:
Pivot view is working :+1: 

> DL Pivot: I miss cost fields (unit price and total)
Changed code as per comments, except the above point as I could not find the mentioned fields.

You are right they must be function (maybe make them storable in both BR and DL): Total procurement / Total tasks

elicoidal commented:
thanks @dreispt for your review

sudhir-serpentcs commented:
@dreispt Have made the requested changes.

@elicoidal And have added Total procurement / Total tasks to the pivot view, yes had to make it storable.

thanks for your review guys!

elicoidal commented:
LGTM too! merging
